packages/squiggle-lang/src/serialization/index.ts

```diff
@@ -0,0 +1,210 @@
+import { JsonValue } from "../utility/typeHelpers.js";
+
+/**
+ * This file contains a generic serialization system.
+ *
+ * Features:
+ * - type-safe
+ * - supports multiple data types
+ * - can serialize any DAG (directed acyclic graph) and deduplicate values
+ * - can serialize multiple values in a single bundle, and deserialize them back based on "entrypoints"
+ *
+ * Squiggle-specific adaptation of it is in `./squiggle.ts`.
+ *
+ * Terminology:
+ * - "Entity" is a type of data that can be serialized and deserialized (e.g. "value", "expression" for Squiggle); one multi-type codec can handle multiple entities.
+ * - "Node" is an instance of an entity type. I use "node" instead of "value" to avoid confusion with the "value" entity type.
+ * - "Bundle" is a JSON-serializable object that contains serialized values.
+ * - "Entrypoint" is a reference to a serialized value in a bundle. When you add a value to a bundle, you get an entrypoint that you can use to refer to this value later.
+ * - "Visitor" is an helper object that can serialize or deserialize nested nodes.
+ *
+ * To make things easier to understand, you can experiment with running Squiggle code with `embedded-with-serialization` runner and `PRINT_SERIALIZED_BUNDLE=1` enabled.
+ * It will print a JSON-serialized bundle to the console.
+ * Example: `PRINT_SERIALIZED_BUNDLE=1 pnpm --silent cli run -r embedded-with-serialization -e 'x=2+2;y=[x,x];z()=x'`
```

Comment:
Coming up with a minimal example is hard! Even this example doesn't show all benefits of deduplication, and its output might be hard to decipher.

Things to keep in mind:
- serialized entities refer to nested entities by id (just by id! not by type, since deserializers already know what type should be there)
- bundle in `embedded-with-serialization` includes result+bindings+exports; so it will include more values than there are defined in the model (you can try `-e '1'` for the minimal example)